File: app/src/main/java/me/ykrank/s1next/data/db/dbmodel/BlackList.java
package me.ykrank.s1next.data.db.dbmodel;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.text.TextUtils;

import org.greenrobot.greendao.annotation.Entity;
import org.greenrobot.greendao.annotation.Generated;
import org.greenrobot.greendao.annotation.Id;
import org.greenrobot.greendao.annotation.Index;
import org.greenrobot.greendao.annotation.Property;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import me.ykrank.s1next.R;

/**
 * 黑名单数据库
 * 使用了ActiveAndroid
 * Created by AdminYkrank on 2016/2/23.
 */
@Entity(nameInDb = "BlackList")
public class BlackList implements Parcelable {
    public static final Creator<BlackList> CREATOR = new Creator<BlackList>() {
        @Override
        public BlackList createFromParcel(Parcel in) {
            return new BlackList(in);
        }

        @Override
        public BlackList[] newArray(int size) {
            return new BlackList[size];
        }
    };
    public static final int NORMAL = 0;
    public static final int HIDE_POST = 1;
    public static final int DEL_POST = 2;
    public static final int HIDE_FORUM = 3;
    public static final int DEL_FORUM = 4;
    private static final String TimeFormat = "yyyy-MM-dd HH:mm";

    @Id(autoincrement = true)
    @Nullable
    private Long id;

    /**
     * Id
     */
    @Property(nameInDb = "AuthorId")
    @Index(name = "IDX_BlackList_AuthorId")
    private int authorId;
    /**
     * 用户名
     */
    @Property(nameInDb = "Author")
    @Index(name = "IDX_BlackList_Author")
    private String author;
    /**
     * 回复的屏蔽状态
     */
    @Property(nameInDb = "Post")
    @PostFLag
    private int post = NORMAL;
    /**
     * 主题的屏蔽状态
     */
    @Property(nameInDb = "Forum")
    @ForumFLag
    private int forum = NORMAL;
    /**
     * 屏蔽时的备注
     */
    @Property(nameInDb = "Remark")
    private String remark;
    /**
     * 屏蔽时的时间
     */
    @Property(nameInDb = "Timestamp")
    private long timestamp;
    /**
     * 是否已同步
     */
    @Property(nameInDb = "Upload")
    private boolean upload;

    @SuppressWarnings("WrongConstant")
    protected BlackList(Parcel in) {
        boolean hasId = in.readByte() == 1;
        if (hasId) {
            id = in.readLong();
        }
        authorId = in.readInt();
        author = in.readString();
        post = in.readInt();
        forum = in.readInt();
        remark = in.readString();
        timestamp = in.readLong();
        upload = in.readByte() != 0;
    }

    public BlackList() {
        this.timestamp = System.currentTimeMillis();
    }

    public BlackList(int authorId, String name, @PostFLag int post, @ForumFLag int forum) {
        this.authorId = authorId;
        this.author = name;
        this.post = post;
        this.forum = forum;
        this.remark = "";
        this.timestamp = System.currentTimeMillis();
        this.upload = false;
    }

    @Generated(hash = 536415208)
    public BlackList(Long id, int authorId, String author, int post, int forum, String remark,
                     long timestamp, boolean upload) {
        this.id = id;
        this.authorId = authorId;
        this.author = author;
        this.post = post;
        this.forum = forum;
        this.remark = remark;
        this.timestamp = timestamp;
        this.upload = upload;
    }

    public void mergeFrom(BlackList bList) {
        if (bList.authorId > 0) {
            this.authorId = bList.authorId;
        }
        if (!TextUtils.isEmpty(bList.author)) {
            this.author = bList.author;
        }
        this.post = bList.post;
        this.forum = bList.forum;
        if (!TextUtils.isEmpty(bList.remark)) {
            this.remark = bList.remark;
        }
        this.timestamp = bList.timestamp;
        this.upload = bList.upload;
    }

    @Nullable
    public Long getId() {
        return id;
    }

    public void setId(@Nullable Long id) {
        this.id = id;
    }

    public static String getTimeFormat() {
        return TimeFormat;
    }

    public int getAuthorId() {
        return authorId;
    }

    public void setAuthorId(int authorId) {
        this.authorId = authorId;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    @PostFLag
    public int getPost() {
        return post;
    }

    public void setPost(@PostFLag int post) {
        this.post = post;
    }

    @ForumFLag
    public int getForum() {
        return forum;
    }

    public void setForum(@ForumFLag int forum) {
        this.forum = forum;
    }

    @Nullable
    public String getRemark() {
        return remark;
    }

    public void setRemark(String remark) {
        this.remark = remark;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public boolean isUpload() {
        return upload;
    }

    public void setUpload(boolean upload) {
        this.upload = upload;
    }

    @StringRes
    public int getPostRes() {
        switch (post) {
            case HIDE_POST:
                return R.string.blacklist_flag_hide;
            case DEL_POST:
                return R.string.blacklist_flag_del;
            default:
                return R.string.blacklist_flag_normal;
        }
    }

    @StringRes
    public int getForumRes() {
        switch (forum) {
            case HIDE_FORUM:
                return R.string.blacklist_flag_hide;
            case DEL_FORUM:
                return R.string.blacklist_flag_del;
            default:
                return R.string.blacklist_flag_normal;
        }
    }

    public boolean isForumHide() {
        return forum == HIDE_FORUM;
    }

    public boolean isPostHide() {
        return post == HIDE_POST;
    }

    public String getTime() {
        SimpleDateFormat sdf = new SimpleDateFormat(TimeFormat, Locale.getDefault());
        return sdf.format(new Date(timestamp));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        if (id == null) {
            dest.writeByte((byte) 0);
        } else {
            dest.writeByte((byte) 1);
            dest.writeLong(id);
        }
        dest.writeInt(authorId);
        dest.writeString(author);
        dest.writeInt(post);
        dest.writeInt(forum);
        dest.writeString(remark);
        dest.writeLong(timestamp);
        dest.writeByte((byte) (upload ? 1 : 0));
    }

    @Override
    public String toString() {
        return "BlackList{" +
                "authorId=" + authorId +
                ", author='" + author + '\'' +
                ", post=" + post +
                ", forum=" + forum +
                ", remark='" + remark + '\'' +
                ", timestamp=" + getTime() +
                ", upload=" + upload +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        BlackList blackList = (BlackList) o;

        if (authorId != blackList.authorId) return false;
        if (post != blackList.post) return false;
        if (forum != blackList.forum) return false;
        if (timestamp != blackList.timestamp) return false;
        if (upload != blackList.upload) return false;
        if (id != null ? !id.equals(blackList.id) : blackList.id != null) return false;
        if (author != null ? !author.equals(blackList.author) : blackList.author != null)
            return false;
        return remark != null ? remark.equals(blackList.remark) : blackList.remark == null;
    }

    @Override
    public int hashCode() {
        int result = id != null ? id.hashCode() : 0;
        result = 31 * result + authorId;
        result = 31 * result + (author != null ? author.hashCode() : 0);
        result = 31 * result + post;
        result = 31 * result + forum;
        result = 31 * result + (remark != null ? remark.hashCode() : 0);
        result = 31 * result + (int) (timestamp ^ (timestamp >>> 32));
        result = 31 * result + (upload ? 1 : 0);
        return result;
    }

    public boolean getUpload() {
        return this.upload;
    }

    public void setId(long id) {
        this.id = id;
    }

    @IntDef({NORMAL, HIDE_POST, DEL_POST})
    public @interface PostFLag {

    }

    @IntDef({NORMAL, HIDE_FORUM, DEL_FORUM})
    public @interface ForumFLag {

    }
}

File: app/src/main/java/me/ykrank/s1next/data/db/dbmodel/BlackList.kt
package me.ykrank.s1next.data.db.dbmodel

import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import me.ykrank.s1next.R
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * 黑名单数据库
 * 使用了ActiveAndroid
 * Created by AdminYkrank on 2016/2/23.
 */
@Entity(
    tableName = "BlackList",
    indices = [
        Index(value = ["AuthorId"], name = "IDX_BlackList_AuthorId"),
        Index(value = ["Author"], name = "IDX_BlackList_Author"),
    ]
)
class BlackList : Parcelable {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "_id")
    var id: Long? = null

    /**
     * Id
     */
    @ColumnInfo(name = "AuthorId")
    var authorId = 0

    /**
     * 用户名
     */
    @ColumnInfo(name = "Author")
    var author: String? = null

    /**
     * 回复的屏蔽状态
     */
    @get:PostFLag
    @ColumnInfo(name = "Post")
    @PostFLag
    var post = NORMAL

    /**
     * 主题的屏蔽状态
     */
    @get:ForumFLag
    @ColumnInfo(name = "Forum")
    @ForumFLag
    var forum = NORMAL

    /**
     * 屏蔽时的备注
     */
    @ColumnInfo(name = "Remark")
    var remark: String? = null

    /**
     * 屏蔽时的时间
     */
    @ColumnInfo(name = "Timestamp")
    var timestamp: Long

    /**
     * 是否已同步
     */
    @ColumnInfo(name = "Upload")
    var isUpload = false

    constructor(`in`: Parcel) {
        val hasId = `in`.readByte().toInt() == 1
        if (hasId) {
            id = `in`.readLong()
        }
        authorId = `in`.readInt()
        author = `in`.readString()
        post = `in`.readInt()
        forum = `in`.readInt()
        remark = `in`.readString()
        timestamp = `in`.readLong()
        isUpload = `in`.readByte().toInt() != 0
    }

    constructor() {
        timestamp = System.currentTimeMillis()
    }

    constructor(authorId: Int, name: String?, @PostFLag post: Int, @ForumFLag forum: Int) {
        this.authorId = authorId
        author = name
        this.post = post
        this.forum = forum
        remark = ""
        timestamp = System.currentTimeMillis()
        isUpload = false
    }

    constructor(
        id: Long?, authorId: Int, author: String?, post: Int, forum: Int, remark: String?,
        timestamp: Long, upload: Boolean
    ) {
        this.id = id
        this.authorId = authorId
        this.author = author
        this.post = post
        this.forum = forum
        this.remark = remark
        this.timestamp = timestamp
        isUpload = upload
    }

    fun mergeFrom(bList: BlackList) {
        if (bList.authorId > 0) {
            authorId = bList.authorId
        }
        if (!bList.author.isNullOrEmpty()) {
            author = bList.author
        }
        post = bList.post
        forum = bList.forum
        if (!bList.remark.isNullOrEmpty()) {
            remark = bList.remark
        }
        timestamp = bList.timestamp
        isUpload = bList.isUpload
    }

    @get:StringRes
    val postRes: Int
        get() = when (post) {
            HIDE_POST -> R.string.blacklist_flag_hide
            DEL_POST -> R.string.blacklist_flag_del
            else -> R.string.blacklist_flag_normal
        }

    @get:StringRes
    val forumRes: Int
        get() = when (forum) {
            HIDE_FORUM -> R.string.blacklist_flag_hide
            DEL_FORUM -> R.string.blacklist_flag_del
            else -> R.string.blacklist_flag_normal
        }
    val isForumHide: Boolean
        get() = forum == HIDE_FORUM
    val isPostHide: Boolean
        get() = post == HIDE_POST
    val time: String
        get() {
            val sdf = SimpleDateFormat(timeFormat, Locale.getDefault())
            return sdf.format(Date(timestamp))
        }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        id.apply {
            if (this == null) {
                dest.writeByte(0.toByte())
            } else {
                dest.writeByte(1.toByte())
                dest.writeLong(this)
            }
        }
        dest.writeInt(authorId)
        dest.writeString(author)
        dest.writeInt(post)
        dest.writeInt(forum)
        dest.writeString(remark)
        dest.writeLong(timestamp)
        dest.writeByte((if (isUpload) 1 else 0).toByte())
    }

    override fun toString(): String {
        return "BlackList{" +
                "authorId=" + authorId +
                ", author='" + author + '\'' +
                ", post=" + post +
                ", forum=" + forum +
                ", remark='" + remark + '\'' +
                ", timestamp=" + time +
                ", upload=" + isUpload +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val blackList = other as BlackList
        if (authorId != blackList.authorId) return false
        if (post != blackList.post) return false
        if (forum != blackList.forum) return false
        if (timestamp != blackList.timestamp) return false
        if (isUpload != blackList.isUpload) return false
        if (if (id != null) id != blackList.id else blackList.id != null) return false
        if (if (author != null) author != blackList.author else blackList.author != null) return false
        return if (remark != null) remark == blackList.remark else blackList.remark == null
    }

    override fun hashCode(): Int {
        var result = if (id != null) id.hashCode() else 0
        result = 31 * result + authorId
        result = 31 * result + if (author != null) author.hashCode() else 0
        result = 31 * result + post
        result = 31 * result + forum
        result = 31 * result + if (remark != null) remark.hashCode() else 0
        result = 31 * result + (timestamp xor (timestamp ushr 32)).toInt()
        result = 31 * result + if (isUpload) 1 else 0
        return result
    }

    @IntDef(NORMAL, HIDE_POST, DEL_POST)
    annotation class PostFLag

    @IntDef(NORMAL, HIDE_FORUM, DEL_FORUM)
    annotation class ForumFLag
    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<BlackList?> = object : Parcelable.Creator<BlackList?> {
            override fun createFromParcel(`in`: Parcel): BlackList {
                return BlackList(`in`)
            }

            override fun newArray(size: Int): Array<BlackList?> {
                return arrayOfNulls(size)
            }
        }
        const val NORMAL = 0
        const val HIDE_POST = 1
        const val DEL_POST = 2
        const val HIDE_FORUM = 3
        const val DEL_FORUM = 4
        const val timeFormat = "yyyy-MM-dd HH:mm"
    }
}